clang/include/clang/Gold/GoldDeclarator.h

```diff
@@ -0,0 +1,448 @@
+
+  /// Does not match the syntactic form of a declarator.
+  DK_Unknown,
+
+  /// The id of a declarator.
+  DK_Identifier,
+
+  /// Declares function parameters.
+  DK_Function,
+
+  /// Declares a type.
+  /// This is whatever is on the RHS of a :
+  /// For example, x:int
+  /// int is labeled as a DK_Type.
+  DK_Type,
+
+  /// This indicates the global namespace specifier was used to name the
+  /// declaration.
+  DK_GlobalNamespecifier,
+
+  /// This is for things which are defined outside of their declared scope.
+  /// Example: a.b. The name specifier would be a
+  DK_NestedNameSpecifier,
+
+  /// Template indication for classes. This part of the declarator is used to
+  /// track if a templated type declaration is being given.
+  DK_TemplateParams,
+
+  /// This is a special declarator type that really doesn't do anything
+  /// with the exception of constructing an empty template parameter list
+  /// This occurs only during explicit specialization x[^int]
+  DK_ImplicitEmptyTemplateParams,
+
+  /// This is for when we have x[^int] or x[T:type][^T]
+  DK_Specialization,
+
+  /// This declarator indicates that there was an error evaluating
+  /// the declarator. This usually means that there is an ErrorSyntax node
+  /// located within the tree.
+  DK_Error,
+};
+
+/// The attribute collection type.
+using Attributes = llvm::SmallVector<const Syntax *, 16>;
+
+/// The structure associated with derived class, by tag name.
+/// DK_Unknown
+///   - has pointer to Syntax node indicating which node was unknown.
+///
+/// DK_Error
+///   - has pointer to Syntax node indicating the location of the error node.
+///
+/// DK_GlobalNamespecifier
+///   - Has pointer to the SyntaxCall containing the "operator'.'".
+///
+/// DK_Identifier
+///   - Has pointer to AtomSyntax indicating the final part of the name.
+///
+/// DK_NestedNameSpecifier
+///   - Has Has pointer AtomSyntax indicate the nested name specifier
+///
+/// DK_Function
+///   - Has a pointer to a Syntax node that contains the function parameter
+///   - gold::Scope - A pointer to the scope that the parameter were
+///     constructed within.
+///
+/// DK_Type
+///   - A pointer to the Syntax node for the root of the type expression.
+///
+/// DK_TemplateParams
+///   - A Syntax node pointing to the parameters contained within an element node
+///   - A gold::Scope that contains parameter declarations
+///   - clang::TemplateArgumentListInfo
+///
+/// DK_ImplicitEmptyTemplateParams
+///   - Pointer to syntax node that's associated with an explicit specialization.
+///   - A pointer to a gold::Scope, that is empty.
+///   - clang::TemplateParameterList
+///
+/// DK_Specialization
+///   - A pointer to the node containing the specialization arguments.
+///   - clang::TemplateArgumentListInfo
+///
+/// A declarator introduces the declaration of a value.
+///
+/// TODO: Represent multiple declarators whose syntax would be
+/// something like: x, y : int -- maybe.
+class Declarator {
```

Comment:
Why have a a class hierarchy when we already have a discriminator? The point of the discriminator is to avoid class hierarchies.